zebra-chain/src/transparent/address.rs

```diff
@@ -168,16 +183,40 @@ impl ZcashDeserialize for Address {
     }
 }
 
-trait ToAddressWithNetwork {
+pub trait ToAddressWithNetwork {
     /// Convert `self` to an `Address`, given the current `network`.
     fn to_address(&self, network: Network) -> Address;
 }
 
 impl ToAddressWithNetwork for Script {
+    // https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/legacy.rs#L43
     fn to_address(&self, network: Network) -> Address {
-        Address::PayToScriptHash {
-            network,
-            script_hash: Address::hash_payload(&self.0[..]),
+        if self.0.len() == 25
+            && self.0[0..3] == [OpCode::Dup as u8, OpCode::Hash160 as u8, 0x14]
+            && self.0[23..25] == [OpCode::EqualVerify as u8, OpCode::CheckSig as u8]
+        {
+            let mut script_hash = [0; 20];
+            script_hash.copy_from_slice(&self.0[3..23]);
+
+            Address::PayToScriptHash {
+                network,
+                script_hash,
+            }
+        } else if self.0.len() == 23
+            && self.0[0..2] == [OpCode::Hash160 as u8, 0x14]
+            && self.0[22] == OpCode::Equal as u8
+        {
+            let mut script_hash = [0; 20];
+            script_hash.copy_from_slice(&self.0[2..22]);
+            Address::PayToScriptHash {
+                network,
+                script_hash,
+            }
+        } else {
+            Address::PayToScriptHash {
+                network,
+                script_hash: Address::hash_payload(&self.0[..]),
+            }
```

Comment:
@dconnolly @oxarbitrage in the "no address" case, `zcashd` returns `None`:
https://github.com/zcash/librustzcash/blob/master/zcash_primitives/src/legacy.rs#L58

But Zebra does a hash of the script, just like the public key hash, which seems... weird.

Which is the correct implementation?

Reply:
The code here seems wrong to me. You should follow zcash_primitives.

(The spec doesn't say how to extract the address because the consensus rules never require extracting the address. They only require checking, *given an address*, that the script is of the correct form.)

Reply:
The zcashd seems to be doing this with https://github.com/zcash/zcash/blob/master/src/chainparams.cpp#L792-L802

Reply:
I removed the zcash_primitives usage and created a new algorithm at https://github.com/ZcashFoundation/zebra/pull/1170/commits/e03217c34b2956dbbf659f3574a58f0b2fa83ea0 to validate script form instead of extracting address.

Reply:
That code is not correct: https://github.com/ZcashFoundation/zebra/commit/e03217c34b2956dbbf659f3574a58f0b2fa83ea0#r43966875

Reply:
Thanks @daira. I followed your suggestions at https://github.com/ZcashFoundation/zebra/pull/1170/commits/b592149bc4f8f3d726d0bbd0ca2aabffa562a531. Will be great if you can take a look when you get some time.

Reply:
ACK b592149